### PR TITLE
List replicas in SHOW CLUSTERS

### PR DESCRIPTION
**Pull request by umanwizard**
Add an array of replicas to the output of `SHOW CLUSTERS`. My reasoning for wanting to add this is that it is the most commonly useful information about a cluster. There has been some concern that this makes our `SHOW` commands somewhat "denormalized", since we are now duplicating information between `SHOW CLUSTERS` and `SHOW CLUSTER REPLICAS`. However, my personal opinion is that it doesn't matter, because the `SHOW` commands are meant to be broadly useful to users, not to be as mathematically elegant as possible.

The rationale for which fields to include in the tuples are those that are mandatory when creating a replica as part of the `CREATE CLUSTER` command, i.e., name and size.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
* Add replica sizes and names to `SHOW CLUSTERS`

### CODE REVIEW

**Comment by rjnn** [comment]
Here to strongly endorse this reasoning:

> However, my personal opinion is that it doesn't matter, because the SHOW commands are meant to be broadly useful to users, not to be as mathematically elegant as possible.

> The rationale for which fields to include in the tuples are those that are mandatory when creating a replica as part of the CREATE CLUSTER command, i.e., name and size.

File: src/sql/src/plan/statement/show.rs [hunk 1]
```diff
@@ -561,8 +561,15 @@ pub fn show_clusters<'a>(
     scx: &'a StatementContext<'a>,
     filter: Option<ShowStatementFilter<Aug>>,
 ) -> Result<ShowSelect<'a>, PlanError> {
-    let query = "SELECT mz_clusters.name FROM mz_catalog.mz_clusters".to_string();
-
+    let query = "
+SELECT
+    mz_clusters.name,
+    pg_catalog.array_agg(ROW(mz_cluster_replicas.name, mz_cluster_replicas.size) ORDER BY mz_cluster_replicas.name)
```
Comment (mjibson) [inline src/sql/src/plan/statement/show.rs:567]: Thoughts on making this json instead? That'd allow us to name the fields, easily adding or removing some later on.

**Comment by benesch** [comment]
> There has been some concern that this makes our `SHOW` commands somewhat "denormalized", since we are now duplicating information between `SHOW CLUSTERS` and `SHOW CLUSTER REPLICAS`. However, my personal opinion is that it doesn't matter, because the `SHOW` commands are meant to be broadly useful to users, not to be as mathematically elegant as possible.

Sorry, I didn't communicate clearly! Here's what I wrote on Slack:

> How would we explain to users when they should use SHOW CLUSTERS vs SHOW CLUSTER REPLICAS? I’m not sure we have precedent for SHOW commands that show information about something other than the thing you are asking to be shown.

My concern wasn't about the denormalization (denormalization is _good_ in `SHOW` commands), but rather how do we literally explain the difference in the docs? If the description of `SHOW CLUSTERS` is "displays information about clusters and their replicas" then what is the description of `SHOW CLUSTER REPLICAS`?

After chatting live, we started to circle around a good justification for what is important enough in `SHOW CLUSTERS`:

> The rationale for which fields to include in the tuples are those that are mandatory when creating a replica as part of the CREATE CLUSTER command, i.e., name and size.

This sort of philosophy makes writing the docs easier, and also makes it easier to determine whether future replica options belong in `SHOW CLUSTERS` or not.

Also, @frankmcsherry has been pushing for more design to take place (cc @parkerhendo) and I think this is a good moment to try to exercise that muscle, rather than rushing to add something in response to one user's (@rjnn's) frustration in one moment.

In particular, I find the tuple formatting to be fairly unreadable! The `SHOW` commands are meant to be almost entirely for humans looking at their terminals, and not for machines, so I'd prefer to go all in on that and use nice text formatting to render the list of replicas. E.g.:

```
> SHOW CLUSTERS;
 cluster |        replicas         
---------+-------------------------
 prod    | r1 (small), r2 (xsmall)
 stg     | r1 (small)
(2 rows)
```

```
> SHOW CLUSTERS;
 cluster |  replicas   
---------+-------------
 prod    | r1 (small) +
         | r2 (xsmall)
 stg     | r1 (small)
(2 rows)
```

```
> SHOW CLUSTERS;
 cluster |  replicas   
---------+-------------
 prod    | r1 (size: small) +
         | r2 (size: xsmall)
 stg     | r1 (size: small)
(2 rows)
```

Thoughts?

**Comment by parkerhendo** [comment]
> The rationale for which fields to include in the tuples are those that are mandatory when creating a replica as part of the CREATE CLUSTER command, i.e., name and size.

I'd **+1** this.

> In particular, I find the tuple formatting to be fairly unreadable! The SHOW commands are meant to be almost entirely for humans looking at their terminals and not for machines, so I'd prefer to go all in on that and use nice text formatting to render the list of replicas. E.g.:
>

> ```
> SHOW CLUSTERS;
>  cluster |        replicas         
> ---------+-------------------------
>  prod    | r1 (small), r2 (xsmall)
>  stg     | r1 (small)
> (2 rows)
> ```
...and something like this.

One thing I would say is that as a user, I would expect the `SHOW CLUSTERS` command to display the same information I see on the `/clusters` page in the web console. This would apply to all objects we display in consolidated lists (Sources, Sinks, Connections, etc.). Right now, we show the **number of replicas** for each cluster in the console, but I'm not 100% sure that's helpful or still the right thing to do.

File: test/sqllogictest/cluster.slt [hunk 2]
```diff
@@ -93,17 +93,17 @@ u1  default
...
-mz_system
-mz_introspection
-bar
-default
-foo
+bar  {"(r1,1)","(r2,1)"}
+default  {"(r1,1)"}
+foo  {"(r1,)"}
+mz_introspection  {"(r1,1)"}
+mz_system  {"(r1,1)"}
```
Comment (benesch) [inline test/sqllogictest/cluster.slt:103]: This is not yet updated for your latest commit, right?

**Review by benesch — APPROVED** [review]
LGTM once you get the tests passing, though please have someone from surfaces sign off too!

File: src/sql/src/plan/statement/show.rs [hunk 3]
```diff
@@ -595,13 +595,24 @@ pub fn show_columns<'a>(
...
     scx: &'a StatementContext<'a>,
     filter: Option<ShowStatementFilter<Aug>>,
 ) -> Result<ShowSelect<'a>, PlanError> {
-    let query = "SELECT mz_clusters.name FROM mz_catalog.mz_clusters".to_string();
-
-    ShowSelect::new(scx, query, filter, None, Some(&["name"]))
+    let query = "
+SELECT
+    mc.name,
+    pg_catalog.string_agg(mcr.name || ' (' || mcr.size || ')', ', ' ORDER BY mcr.name)
```
Comment (mjibson) [inline src/sql/src/plan/statement/show.rs:608]: I still think this would be better as a json array than a string. Cluster names with commas or parens could make this hard to parse, whereas if this were a json array of objects with fields name and size, it becomes trivial.

Comment (benesch) [reply src/sql/src/plan/statement/show.rs:608]: Hard to parse for humans or for computers? @parkerhendo and I were coalescing above around a philosophy that `SHOW` commands are meant for humans staring at their terminals, not computers. (Computers already have a great interface, in the highly normalized schema in `mz_catalog`.)

For me, using JSON arrays doesn't pass the readability test for humans!

```
SHOW REPLICAS;
 cluster |                               replicas                               
---------+----------------------------------------------------------------------
 c1      | [{"name": "r1", "size": "small"}, {"name": "r2", "size": "2xsmall"}]
 c2      | [{"name": "r2", "size": "medium"}]
```

Comment (mjibson) [reply src/sql/src/plan/statement/show.rs:608]: SGTM.

**Review by mjibson — APPROVED** [review]
